pkg/controllers/management/nodetemplate/nodetemplate_handler.go

```diff
@@ -0,0 +1,294 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+)
+
+const (
+	ownerBindingsAnno = "ownerBindingsCreated"
+)
+
+type nodeTemplateController struct {
+	roleLister v3.GlobalRoleLister
+	roleClient v3.GlobalRoleInterface
+	rbLister   v3.GlobalRoleBindingLister
+	rbClient   v3.GlobalRoleBindingInterface
+	ntClient   v3.NodeTemplateInterface
+	ntLister   v3.NodeTemplateLister
+	npLister   v3.NodePoolLister
+	npClient   v3.NodePoolInterface
+	nsLister   v1.NamespaceLister
+	nsClient   v1.NamespaceInterface
+	mgmtCtx    *config.ManagementContext
+}
+
+func Register(ctx context.Context, mgmt *config.ManagementContext) {
+	nt := nodeTemplateController{
+		roleLister: mgmt.Management.GlobalRoles("").Controller().Lister(),
+		roleClient: mgmt.Management.GlobalRoles(""),
+		rbLister:   mgmt.Management.GlobalRoleBindings("").Controller().Lister(),
+		rbClient:   mgmt.Management.GlobalRoleBindings(""),
+		ntClient:   mgmt.Management.NodeTemplates(""),
+		ntLister:   mgmt.Management.NodeTemplates("").Controller().Lister(),
+		npLister:   mgmt.Management.NodePools("").Controller().Lister(),
+		npClient:   mgmt.Management.NodePools(""),
+		nsLister:   mgmt.Core.Namespaces("").Controller().Lister(),
+		nsClient:   mgmt.Core.Namespaces(""),
+		mgmtCtx:    mgmt,
+	}
+
+	mgmt.Management.NodeTemplates("").Controller().AddHandler(ctx, "nt-grb-handler", nt.sync)
+}
+
+func (nt *nodeTemplateController) sync(key string, nodeTemplate *v3.NodeTemplate) (runtime.Object, error) {
+	if nodeTemplate == nil || nodeTemplate.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	metaAccessor, err := meta.Accessor(nodeTemplate)
+	if err != nil {
+		return nodeTemplate, err
+	}
+
+	creatorID, ok := metaAccessor.GetAnnotations()[rbac.CreatorIDAnn]
+	if !ok {
+		return nodeTemplate, fmt.Errorf("clusterTemplate %v has no creatorId annotation", metaAccessor.GetName())
+	}
+
+	// Parts of the node template in k8s are dynamic, such as an azureConfig. If the dynamic client is not use then the
+	// dynamic fields, like azureConfig, are stripped when marshalling the data into the NodeTemplate struct. If changes
+	// are made, and that struct is committed to k8s, then the dynamic field is effectively removed.
+	restConfig := nt.mgmtCtx.RESTConfig
+	dynamicClient, err := dynamic.NewForConfig(&restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	s := schema.GroupVersionResource{
+		Group:    "management.cattle.io",
+		Version:  "v3",
+		Resource: "nodetemplates",
+	}
+	ntDynamicClient := dynamicClient.Resource(s)
+
+	if nodeTemplate.Namespace != namespace.NodeTemplateGlobalNamespace {
+		// node template must be migrated to global namespace
+		if err := nt.migrateNodeTemplate(ntDynamicClient, nodeTemplate); err != nil {
+			return nil, err
+		}
+
+		// node template has been deleted- handler should return nil object
+		nodeTemplate = nil
+	} else {
```

Comment:
Drop the else

Reply:
The migration should give back a new version of the nodetemplate in the global ns, use that one to create the bindings and update the annotation once completed